sklearn/metrics/_pairwise_distances_reduction/_argkminlabels.pyx

```diff
@@ -0,0 +1,203 @@
+        chunk_size=None,
+        strategy=None,
+        ITYPE_t k=1,
+        weights=None,
+        cnp.ndarray[ndim=1, dtype=ITYPE_t] labels=None,
+    ):
+        super().__init__(
+            datasets_pair=datasets_pair,
+            chunk_size=chunk_size,
+            strategy=strategy,
+            k=k,
+        )
+
+        if weights == "uniform":
+            self.weight_type = Weight.uniform
+        elif weights == "distance":
+            self.weight_type = Weight.distance
+        else:
+            self.weight_type = Weight.other
+        self.labels = labels
+        self.unique_labels = np.unique(labels)
+
+        # Heap to be returned
+        self.argkmin_labels = np.full(self.n_samples_X, -1, dtype=ITYPE)
+
+        # Map from set of unique labels to their indices in `label_counts`
+        self.labels_to_index = {label:idx for idx, label in enumerate(self.unique_labels)}
+        # Buffer used in building a histogram for one-pass weighted mode
+        self.label_counts = np.zeros((self.n_samples_X,  len(self.unique_labels)), dtype=DTYPE)
+
+    def _finalize_results(self, bint return_distance=False):
+        if return_distance:
+            # We need to recompute distances because we relied on
+            # surrogate distances for the reduction.
+            self.compute_exact_distances()
+
+            # Values are returned identically to the way `KNeighborsMixin.kneighbors`
+            # returns values. This is counter-intuitive but this allows not using
+            # complex adaptations where `PairwiseDistancesArgKmin64.compute` is called.
+            return np.asarray(self.argkmin_distances), np.asarray(self.argkmin_indices), np.asarray(self.argkmin_labels)
+
+        return np.asarray(self.argkmin_indices), np.asarray(self.argkmin_labels)
+
+    cdef void _parallel_on_X_prange_iter_finalize(
+        self,
+        ITYPE_t thread_num,
+        ITYPE_t X_start,
+        ITYPE_t X_end,
+    ) nogil:
+        cdef:
+            ITYPE_t idx, jdx, y_idx, label_index, sample_index
+            DTYPE_t max_weight
+            ITYPE_t label, max_label
+            DTYPE_t val, weight
+        # Sorting the main heaps portion associated to `X[X_start:X_end]`
+        # in ascending order w.r.t the distances.
+        for idx in range(X_end - X_start):
+            simultaneous_sort(
+                self.heaps_r_distances_chunks[thread_num] + idx * self.k,
+                self.heaps_indices_chunks[thread_num] + idx * self.k,
+                self.k
+            )
+            # One-pass top-one weighted mode
+            sample_index = thread_num * self.X_n_samples_chunk + idx
+            max_weight = -1
+            for jdx in range(self.k): # Iterate through k-nearest neighbors
+                if self.weight_type == Weight.uniform:
+                    weight = 1.
+                elif self.weight_type == Weight.distance:
+                    weight = 1. / self.heaps_r_distances_chunks[thread_num][idx*self.k+jdx]
```

Comment:
Having branching here might be costly.

Eventually, we might prefer one of the options (not exhaustive):
 - have it before the looping with code duplication (harder for maintenance)
 - come up with a branchless computation of `weight` using the underlying value of `self.weight_type` (efficient, but involve a hacky one-liner and harden introduction of new value for `Weight` generally)
 - wrap this logic via a template method and thus subclass this class for each strategy (probably nicer but also heavy regarding the flexibility of OOP in Cython)

What do you think?

Reply:
I think it may be somewhat negligible here. I tried out three configurations:
1. The current one, denoted as `PR`
2. A modification making use of fused types as a templating strategy (shout-out to @jakirkham for introducing me to this trick via the `isfinite` implementation), denoted as `PR_fused`
3. Entirely branchless by just removing the branch all together and defaulting to uniform, denoted as `PR_control`.

[Script for reference](https://gist.github.com/Micky774/ae74488c832c65a87391e1508ad3c00a).

<details>
<summary> Plot of results </summary>

![077c62c2-81cd-4621-943b-87a2a5b05671](https://user-images.githubusercontent.com/34613774/182915863-c038d5a6-d50d-4def-ba6d-e2347eb112f7.png)

</details>